File: ws01_plumbing/src/cpp01_topic/src/demo01_talker_str.cpp
/*
需求：以某个固定频率发送文本“Hello world!”。文本后缀编号，每发送一条消息，编号递增1。
步骤：
    1.包含头文件；
    2.初始化 ROS2 客户端；
    3.定义节点类：
        3-1.创建发布方；
        3-2.创建定时器；
        3-3.组织消息并发布；
    4.调用spin函数，并传入节点对象指针；
    5.释放资源

*/

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class Talker: public rclcpp::Node
{
public:
    Talker():Node("talker_node_cpp"), count_(0)
    {
        // 3-1.创建发布方
        publisher_ = this->create_publisher<std_msgs::msg::String>("chartter", 10);
        /*
        create_publisher
        模板：被发布的消息类型
        参数：1.话题名称 2.QOS（消息队列长度）
        返回值：发布对象指针
        */

        // 3-2.创建定时器
        timer_ = this->create_wall_timer(1s, std::bind(&Talker::on_timer, this));
        /*
        create_wall_timer
        参数：1.时间间隔 2.回调函数
        返回值：定时器对象指针
        */
    }

private:
    void on_timer()
    {
        // 3-3.组织并发布消息
        auto message = std_msgs::msg::String();
        message.data = "Hello world!" + std::to_string(count_++);
        RCLCPP_INFO(this->get_logger(), "发布的消息： '%s'", message.data.c_str());
        publisher_->publish(message);
    }
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    size_t count_;
};

int main(int argc, char *argv[])
{
    // 2.初始化 ROS2 客户端
    rclcpp::init(argc, argv);

    // 4.调用spin函数，并传入节点对象指针
    rclcpp::spin(std::make_shared<Talker>());

    // 5.释放资源
    rclcpp::shutdown();
    
    return 0;
}

File: ws01_plumbing/src/py01_topic/py01_topic/demo01_talker_str.py
"""
需求：以某个固定频率发送文本"Hello world!"。文本后缀编号,每发送一条消息,编号递增1。
步骤:
    1.导包
    2.初始化 ROS2 客户端
    3.定义节点类
        3-1.创建消息发布方
        3-2.创建定时器
        3-3.组织消息并发布
    4.调用spin函数,并传入节点对象
    5.释放资源
"""

# 导包
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

# 3.自定义节点类
class Talker(Node):
    def __init__(self):
        super().__init__("talker_node_py")

        self.get_logger().info("发布方创建了...")

        # 设置计数器
        self.count = 0

        # 3-1.创建消息发布方
        self.publisher = self.create_publisher(String, "chatter", 10)

        # 3-2.创建定时器
        self.timer = self.create_timer(1.0, self.on_timer)

    def on_timer(self):
        # 3-3.组织消息并发布
        message = String()
        message.data = f"Hello World!{str(self.count)}"
        self.publisher.publish(message)
        self.count += 1
        self.get_logger().info(f"发布的数据:{message.data}")


def main(args=None):

    # 2.初始化 ROS2 客户端
    rclpy.init(args=args)

    # 4.调用spin函数
    rclpy.spin(Talker())

    # 5.释放资源
    rclpy.shutdown()


if __name__ == '__main__':
    main()
